File: Regression/Polynomial Regression/Polynomial Regression.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import train_test_split


# import the data set
dataset = pd.read_csv("Position_Salaries.csv")

# make sure X is a matrix not a vector
x = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Fitting linear model
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(x, y)


# Fitting polynomial model
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree=5)
x_poly = poly_reg.fit_transform(x)
lin_reg2 = LinearRegression()
lin_reg2.fit(x_poly, y)

plt.scatter(x, y, color="red")
plt.plot(x, lin_reg.predict(x), color="blue")
plt.title("hi i am a title")
plt.xlabel("years")
plt.ylabel("salary")
plt.show()

x_grid = np.arange(min(x), max(x), 0.1)
x_grid = x_grid.reshape((len(x_grid), 1))

plt.scatter(x, y, color="red")
plt.plot(x_grid, lin_reg2.predict(poly_reg.fit_transform(x_grid)), color="blue")
plt.title("hi i am a title")
plt.xlabel("years")
plt.ylabel("salary")
plt.show()

print(lin_reg.predict([[6.5]]))
print(lin_reg2.predict(poly_reg.fit_transform([[6.5]])))

File: Regression/Polynomial Regression/PolynomialRegression.R


dataset = read.csv("Position_Salaries.csv")
dataset = dataset[2:3]

# Creating linear model
lin_reg = lm(formula = Salary ~ Level, data = dataset)

# Creating polynomial model
dataset$Level2 = dataset$Level ^ 2
dataset$Level3 = dataset$Level ^ 3
dataset$Level4 = dataset$Level ^ 4
poly_reg = lm(formula = Salary ~ ., data = dataset)

# Visualising data
# install.packages("ggplot2")
library(ggplot2)
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary), colour = "red") + 
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)), colour = "blue") + 
  ggtitle("I am a title") +
  xlab("years") +
  ylab("salary")

ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary), colour = "red") + 
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)), colour = "blue") + 
  ggtitle("I am a title") +
  xlab("years") +
  ylab("salary")              

x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)

ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary), colour = "red") + 
  geom_line(aes(x = x_grid, y = predict(poly_reg, newdata = data.frame(Level = x_grid, 
                                                                       Level2 = x_grid^2,
                                                                       Level3 = x_grid^3,
                                                                       Level4 = x_grid^4))), colour = "blue") + 
  ggtitle("I am a title") +
  xlab("years") +
  ylab("salary")      

# Predict individual response
y_pred = predict(lin_reg, data.frame(Level = 6.5))
print(y_pred)

y_pred = predict(poly_reg, data.frame(Level = 6.5, Level2 = 6.5^2, Level3 = 6.5^3, Level4 = 6.5^4))
print(y_pred)



